src/Microsoft.Health.Dicom.Functions/Indexing/ReindexOperation.cs

```diff
@@ -0,0 +1,141 @@
+            ITagReindexOperationStore tagOperationStore,
+            IInstanceReindexer instanceReindexer)
+        {
+            EnsureArg.IsNotNull(configOptions, nameof(configOptions));
+            EnsureArg.IsNotNull(tagOperationStore, nameof(tagOperationStore));
+            EnsureArg.IsNotNull(instanceReindexer, nameof(instanceReindexer));
+            EnsureArg.IsNotNull(addExtendedQueryTagService, nameof(addExtendedQueryTagService));
+            _reindexConfig = configOptions.Value.Add;
+            _tagOperationStore = tagOperationStore;
+            _instanceReindexer = instanceReindexer;
+            _addExtendedQueryTagService = addExtendedQueryTagService;
+        }
+
+        /// <summary>
+        ///  The orchestration function to add extended query tags.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(AddExtendedQueryTagsOrchestrationAsync))]
+        public async Task AddExtendedQueryTagsOrchestrationAsync(
+            [OrchestrationTrigger] IDurableOrchestrationContext context,
+            ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<AddExtendedQueryTagEntry>>();
+            var tagEntries = await context.CallActivityAsync<IEnumerable<ExtendedQueryTagStoreEntry>>(nameof(AddExtendedQueryTagAsync), input);
+            await context.CallSubOrchestratorAsync(nameof(ReindexExtendedQueryTagsOrchestrationAsync), context.InstanceId, tagEntries);
+        }
+
+        /// <summary>
+        /// The orchestration function to Reindex Extended Query Tags.
+        /// </summary>
+        /// <param name="context">the context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns></returns>
+        [FunctionName(nameof(ReindexExtendedQueryTagsOrchestrationAsync))]
+        public async Task ReindexExtendedQueryTagsOrchestrationAsync(
+           [OrchestrationTrigger] IDurableOrchestrationContext context,
+           ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<ExtendedQueryTagStoreEntry>>();
+
+
+            // TODO: should start a new orchestraion instead of while loop
+            while (true)
+            {
+                IReadOnlyList<long> watermarks = await context
+                    .CallActivityAsync<IReadOnlyList<long>>(nameof(GetNextWatermarksOfOperationAsync), context.InstanceId);
+                if (watermarks.Count == 0)
+                {
+                    break;
+                }
+
+                IReadOnlyList<ExtendedQueryTagStoreEntry> queryTags = await context
```

Comment:
If we're always querying for the set of tags (to eventually support pause/cancel), should we even be passing them to the orchestration in the first place?

Reply:
Sorry, not get you.  If we passing them to orchestration, who should call  GetExtendedQueryTagsOfOperationAsync to get qualified tags?

Reply:
I mean that the orchestration doesn't need the tags in its input since the orchestration is already querying the tags from SQL directly.

Reply:
that's right, updating code...